### PR TITLE
Rupan/market actor tests

### PR DESCRIPTION
**Pull request by nannick**
**Summary of changes**
Changes introduced in this pull request:
-  Updating some Market Actor Logic
-  Implementing some Market Actor Tests
- Misc updates to fix small issues

TODO : 
5 of remaining Market Actor Tests depend on Storage Miner, will not be able to complete those tests until Storage Miner functionality is available
Clean up Warnings

**Reference issue to close (if applicable)**

Closes 

**Other information and links**

### CODE REVIEW

**Comment by nannick** [comment]
Market Actor Logic got updated and they removed a function called HandleExpiredDeals and added a function called CronTick. I'm not sure if these do they same thing, or if there is a fundamental difference between the two.

**Comment by nannick** [comment]
TODO :

Implement the 5 remaining Market Actor Tests that depend on Miner Actor Functionality. They should be ready to go, just a few small changes. Leaving them as fails right now, so the PR can't be mistakenly added

Implement serialize functionality for VerifyDealsOnSectorProveCommitReturn. After @timvermeulen's PR comes in, it should be super easy to do. After this happens make small changes to Market Actor methods that use this. 

Despite these things left I believe the rest is ready for review, So I'm updating the label for this PR

File: utils/test_utils/src/chain_structures.rs [hunk 1]
```diff
@@ -7,13 +7,14 @@ use address::Address;
 use blocks::{
     Block, BlockHeader, EPostProof, EPostTicket, FullTipset, Ticket, Tipset, TipsetKeys, TxMeta,
 };
-use chain::TipsetMetadata;
+
```
Comment (austinabell) [inline utils/test_utils/src/chain_structures.rs:10]:
```suggestion
```

File: vm/actor/src/builtin/market/mod.rs [hunk 2]
```diff
@@ -191,7 +191,7 @@ impl Actor {
         }
 
         // All deals should have the same provider so get worker once
-        let provider_raw = &params.deals[0].proposal.provider;
+        let provider_raw = &params.deals[0].proposal.provider.clone();
```
Comment (austinabell) [inline vm/actor/src/builtin/market/mod.rs:194]:
```suggestion
        let provider_raw = &params.deals[0].proposal.provider;
```

File: vm/actor/src/builtin/market/mod.rs [hunk 3]
```diff
@@ -298,6 +298,7 @@ impl Actor {
     {
         rt.validate_immediate_caller_type(std::iter::once(&*MINER_ACTOR_CODE_ID))?;
         let miner_addr = *rt.message().from();
+
```
Comment (austinabell) [inline vm/actor/src/builtin/market/mod.rs:301]:
```suggestion
```

File: vm/actor/src/builtin/market/mod.rs [hunk 4]
```diff
@@ -501,7 +502,7 @@ impl Actor {
             let mut lt = BalanceTable::from_root(rt.store(), &st.locked_table)
                 .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?;
 
-            let mut i = st.last_cron + 1;
+            let mut i = st.last_cron.unwrap() + 1;
```
Comment (austinabell) [inline vm/actor/src/builtin/market/mod.rs:505]: can you put a TODO on this to fix, it's going to be changed with Epoch changes, but broken as is

File: vm/actor/src/builtin/market/state.rs [hunk 5]
```diff
@@ -420,6 +420,7 @@ impl State {
         amount: &TokenAmount,
     ) -> Result<(), String> {
         // Subtract from locked and escrow tables
+
```
Comment (austinabell) [inline vm/actor/src/builtin/market/state.rs:423]:
```suggestion
```

File: vm/actor/src/builtin/market/types.rs [hunk 6]
```diff
@@ -8,6 +8,7 @@ use clock::ChainEpoch;
 use encoding::tuple::*;
 use fil_types::RegisteredSealProof;
 use num_bigint::biguint_ser;
+
```
Comment (austinabell) [inline vm/actor/src/builtin/market/types.rs:11]:
```suggestion
```

File: vm/actor/src/util/optional_epoch.rs [hunk 7]
```diff
@@ -9,6 +9,12 @@ use std::ops::Deref;
 #[derive(PartialEq, Eq, Copy, Clone, Debug)]
 pub struct OptionalEpoch(pub Option<ChainEpoch>);
 
+impl Default for OptionalEpoch {
+    fn default() -> Self {
+        OptionalEpoch(None)
+    }
+}
+
```
Comment (austinabell) [inline vm/actor/src/util/optional_epoch.rs:17]:
```suggestion
#[derive(PartialEq, Eq, Copy, Clone, Debug, Default)]
pub struct OptionalEpoch(pub Option<ChainEpoch>);

```

File: vm/actor/tests/market_actor_test.rs [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+use message::UnsignedMessage;
+use vm::{ExitCode, Serialized, TokenAmount, METHOD_CONSTRUCTOR, METHOD_SEND};
+
+const OWNER_ID: u64 = 101;
+const PROVIDER_ID: u64 = 102;
+const WORKER_ID: u64 = 103;
+const CLIENT_ID: u64 = 104;
+
+//const caller_types_signable : Vec<Cid>= vec![ACCOUNT_ACTOR_CODE_ID.clone(), MULTISIG_ACTOR_CODE_ID.clone()];
+
```
Comment (austinabell) [inline vm/actor/tests/market_actor_test.rs:27]:
```suggestion
```

**Comment by austinabell** [comment]
The submodule change needs to be updated then I'm good with this :)

**Pull request merged**